flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/ProcessOperator.java

```diff
@@ -114,12 +114,22 @@ public long currentWatermark() {
 
 		@Override
 		public void registerProcessingTimeTimer(long time) {
-			throw new UnsupportedOperationException("Setting timers is only supported on a KeyedStream.");
+			throw new UnsupportedOperationException(UNSUPPORTED_REGISTER_TIMER_MSG);
 		}
 
 		@Override
 		public void registerEventTimeTimer(long time) {
-			throw new UnsupportedOperationException("Setting timers is only supported on a KeyedStream.");
+			throw new UnsupportedOperationException(UNSUPPORTED_REGISTER_TIMER_MSG);
+		}
+
+		@Override
+		public void deleteProcessingTimeTimer(long time) {
+			throw new UnsupportedOperationException(UNSUPPORTED_DELETE_TIMER_MSG);
+		}
+
+		@Override
+		public void deleteEventTimeTimer(long time) {
+			throw new UnsupportedOperationException(UNSUPPORTED_DELETE_TIMER_MSG);
 		}
```

Comment:
I know it is out of scope, but why do we give access to the `TimerService` in a non-keyed context? Wouldn't it be better to follow the interface-segregation principle a bit more here and expose the timer service only to keyed operators? Is there a reason why we don't do it like that @aljoscha?

Reply:
The reason is that `ProcessFunction` can be used both on a keyed stream and a non-keyed stream. There is the newer `KeyedProcessFunction`, which should be the only function that exposes timers and state but for historic reasons it's currently like this. 😅

Reply:
Alright, thanks for the clarification. I guess we could change it with Flink 2.0 because it would break the API to change it.

Reply:
may want to add open a JIRA now so we don't forget it.

Reply:
https://issues.apache.org/jira/browse/FLINK-9529